proposals/testing/NNNN-polling-confirmations.md

```diff
@@ -0,0 +1,512 @@
+}
+```
+
+Specifying `duration` or `interval` directly on either new `confirmation`
+function will override any value provided by the relevant trait. Additionally,
+when multiple of these configuration traits are specified, the innermost or
+last trait will be applied.
+
+### Default Polling Configuration
+
+For all polling confirmations, the Testing library will default `duration` to
+1 second, and `interval` to 1 millisecond.
+
+### Platform Availability
+
+Polling confirmations will not be available on platforms that do not support
+Swift Concurrency.
+
+### Duration and Concurrent Execution
+
+It is an unfortunate side effect that directly using the `duration` to determine
+when to stop polling (i.e. `while duration has not elapsed { poll() }`) is
+unreliable in a parallel execution environment. Especially on systems that are
+under-resourced, under very high load, or both - such as CI systems. This is
+especially the case for the Testing library, which, at time of writing, submits
+every test at once to the concurrency system for scheduling. Under this
+environment, with heavily-burdened machines running test suites with a very
+large amount of tests, there is a very real case that a polling confirmation's
+`duration` might elapse before the `body` has had a chance to return even once.
+
+To prevent this, the Testing library will calculate how many times to poll the
+`body`. This is done by dividing the `duration` by the `interval`. For example,
+with the default 1 second duration and 1 millisecond interval, the Testing
+library will poll 1000 times, waiting 1 millisecond between polling attempts.
+This works and is immune to the issues posed by concurrent execution on
+heavily-burdened systems.
+This is also very easy for test authors to understand and predict, even if it is
+not fully accurate - each poll attempt takes some amount of time, even for very
+fast `body` closures. Which means that the real-time duration of a polling
+confirmation will always be longer than the value specified in the `duration`
+argument.
+
+### Usage
+
+These functions can be used with an async test function:
+
+```swift
+@Test func `The aquarium's dolphin nursery works`() async {
+    let subject = Aquarium()
+    Task {
+        await subject.raiseDolphins()
+    }
+    await confirmation(until: .firstPass) {
+        await subject.dolphins.count == 1
+    }
+}
+```
+
+With the definition of `Aquarium` above, the closure will only need to be
+evaluated a few times before it starts returning true. At which point polling
+will end, and no failure will be reported.
+
+Polling will be stopped in the following cases:
+
+- The specified `duration` has elapsed.
+- If the task that started the polling is cancelled.
+- For `PollingStopCondition.firstPass`: The first time the closure returns true
+  or a non-nil value
+- For `PollingStopCondition.stopsPassing`: The first time the closure returns
+  false or nil.
+- The first time the closure throws an error.
```

Comment:
It might not be clear to a reader that this is an "OR" list not "AND". I find it helpful when proposals clarify that kind of thing by adding _"…, or"_ to the second-to-last bullet. For example:

```suggestion
Polling will be stopped when either:

- the specified duration has elapsed,
- the task that started the polling is cancelled,
- the closure returns a value that satisfies the stopping condition, or
- the closure throws an error.
```

(And in that suggestion I shortened the third bullet, instead of reiterating the stopping condition behaviors, since those are specified elsewhere in the document.)